File: impl/python/agentnet_py/markdown.py
from __future__ import annotations

from typing import List

from .cbor import CborError


class MarkdownError(ValueError):
    pass


MAX_HEADING_LEVEL = 6


def canonicalize_markdown_profile(text: str) -> str:
    normalized = text.replace("\r\n", "\n").replace("\r", "\n")
    out: List[str] = []
    blank_run = 0
    in_code = False

    for raw_line in normalized.split("\n"):
        line = raw_line.replace("\t", "    ").rstrip()

        if in_code:
            if _is_fence_line(line):
                out.append("```")
                in_code = False
                blank_run = 0
                continue
            out.append(line)
            blank_run = 0
            continue

        trimmed = line.strip()
        if trimmed == "":
            blank_run += 1
            if blank_run <= 2:
                out.append("")
            continue
        blank_run = 0

        if _is_fence_line(line):
            out.append(_canonical_fence(line))
            in_code = True
            continue

        out.append(_canonicalize_line(trimmed))

    if in_code:
        raise MarkdownError("unterminated code fence")

    return "\n".join(out)


def validate_markdown_profile(text: str) -> None:
    canonical = canonicalize_markdown_profile(text)
    if canonical != text:
        raise MarkdownError("markdown not canonical")


def _canonicalize_line(line: str) -> str:
    if line.startswith("    "):
        raise MarkdownError("indented code blocks not allowed")
    if _contains_html(line):
        raise MarkdownError("html not allowed")
    if "![" in line:
        raise MarkdownError("images not allowed")
    if _looks_like_table_separator(line):
        raise MarkdownError("tables not allowed")

    heading = _canonicalize_heading(line)
    if heading is not None:
        return heading
    blockquote = _canonicalize_blockquote(line)
    if blockquote is not None:
        return blockquote
    hr = _canonicalize_hr(line)
    if hr is not None:
        return hr
    list_line = _canonicalize_list(line)
    if list_line is not None:
        return list_line

    _validate_links(line)
    return line


def _is_fence_line(line: str) -> bool:
    return line.lstrip().startswith("```")


def _canonical_fence(line: str) -> str:
    trimmed = line.lstrip()
    rest = trimmed[3:]
    lang = rest.strip()
    if lang == "":
        return "```"
    if not all(c.isalnum() or c in "-_+." for c in lang):
        raise MarkdownError("invalid code fence language")
    return f"```{lang}"


def _canonicalize_heading(line: str) -> str | None:
    if not line.startswith("#"):
        return None
    level = 0
    for ch in line:
        if ch == "#":
            level += 1
        else:
            break
    if level == 0 or level > MAX_HEADING_LEVEL:
        raise MarkdownError("invalid heading")
    rest = line[level:].lstrip()
    if rest == "":
        raise MarkdownError("empty heading")
    return f"{'#' * level} {rest}"


def _canonicalize_blockquote(line: str) -> str | None:
    if not line.startswith(">"):
        return None
    rest = line[1:].lstrip()
    return f"> {rest}"


def _canonicalize_hr(line: str) -> str | None:
    if line.strip() == "---":
        return "---"
    return None


def _canonicalize_list(line: str) -> str | None:
    indent, rest = _split_indent(line)
    if indent > 3:
        raise MarkdownError("excessive indent")
    if rest.startswith("-"):
        if len(rest) == 1 or not rest[1].isspace():
            raise MarkdownError("invalid list marker")
        after = rest[1:].lstrip()
        if after == "":
            raise MarkdownError("invalid list marker")
        _validate_links(after)
        return f"{' ' * indent}- {after}"
    if rest.startswith("*") or rest.startswith("+"):
        raise MarkdownError("invalid list marker")
    if "." in rest:
        digits, after_dot = rest.split(".", 1)
        if digits.isdigit():
            if digits != "1":
                raise MarkdownError("ordered list must use 1.")
            if after_dot == "" or not after_dot[0].isspace():
                raise MarkdownError("invalid ordered list")
            after = after_dot.lstrip()
            if after == "":
                raise MarkdownError("invalid ordered list")
            _validate_links(after)
            return f"{' ' * indent}1. {after}"
    return None


def _split_indent(line: str) -> tuple[int, str]:
    count = 0
    for ch in line:
        if ch == " ":
            count += 1
        else:
            break
    return count, line[count:]


def _validate_links(line: str) -> None:
    idx = 0
    while True:
        start = line.find("](", idx)
        if start == -1:
            break
        url_start = start + 2
        end = line.find(")", url_start)
        if end == -1:
            raise MarkdownError("invalid link")
        url = line[url_start:end].strip()
        if ":" not in url:
            raise MarkdownError("invalid link")
        scheme = url.split(":", 1)[0].lower()
        if scheme not in {"https", "agentnet", "did"}:
            raise MarkdownError("invalid link scheme")
        idx = end + 1


def _contains_html(line: str) -> bool:
    for i, ch in enumerate(line):
        if ch == "<" and i + 1 < len(line):
            nxt = line[i + 1]
            if nxt.isalpha() or nxt in {"/", "!"}:
                return True
    return False


def _looks_like_table_separator(line: str) -> bool:
    if "|" not in line:
        return False
    trimmed = line.strip()
    if trimmed == "":
        return False
    for ch in trimmed:
        if ch not in "|:- ":
            return False
    return True

File: impl/rust/crates/anetsdk/src/markdown.rs
use crate::Error;

const MAX_HEADING_LEVEL: usize = 6;

pub fn canonicalize_markdown_profile(input: &str) -> Result<String, Error> {
    let mut normalized = input.replace("\r\n", "\n");
    normalized = normalized.replace('\r', "\n");

    let mut out: Vec<String> = Vec::new();
    let mut blank_run = 0usize;
    let mut in_code_block = false;

    for raw_line in normalized.split('\n') {
        let mut line = raw_line.replace('\t', "    ");
        line = line.trim_end().to_string();

        if in_code_block {
            if is_fence_line(&line) {
                out.push("```".to_string());
                in_code_block = false;
                blank_run = 0;
                continue;
            }
            out.push(line);
            blank_run = 0;
            continue;
        }

        let trimmed = line.trim();
        if trimmed.is_empty() {
            blank_run += 1;
            if blank_run <= 2 {
                out.push(String::new());
            }
            continue;
        }
        blank_run = 0;

        if is_fence_line(&line) {
            let fence = canonical_fence(&line)?;
            out.push(fence);
            in_code_block = true;
            continue;
        }

        let canonical = canonicalize_line(trimmed)?;
        out.push(canonical);
    }

    if in_code_block {
        return Err(Error::Cbor("unterminated code fence"));
    }

    Ok(out.join("\n"))
}

pub fn validate_markdown_profile(input: &str) -> Result<(), Error> {
    let canonical = canonicalize_markdown_profile(input)?;
    if canonical != input {
        return Err(Error::Cbor("markdown not canonical"));
    }
    Ok(())
}

fn canonicalize_line(line: &str) -> Result<String, Error> {
    if line.starts_with("    ") {
        return Err(Error::Cbor("indented code blocks not allowed"));
    }
    if contains_html(line) {
        return Err(Error::Cbor("html not allowed"));
    }
    if line.contains("![") {
        return Err(Error::Cbor("images not allowed"));
    }
    if looks_like_table_separator(line) {
        return Err(Error::Cbor("tables not allowed"));
    }

    if let Some(canonical) = canonicalize_heading(line)? {
        return Ok(canonical);
    }
    if let Some(canonical) = canonicalize_blockquote(line)? {
        return Ok(canonical);
    }
    if let Some(canonical) = canonicalize_hr(line)? {
        return Ok(canonical);
    }
    if let Some(canonical) = canonicalize_list(line)? {
        return Ok(canonical);
    }

    validate_links(line)?;
    Ok(line.to_string())
}

fn is_fence_line(line: &str) -> bool {
    let trimmed = line.trim_start();
    trimmed.starts_with("```")
}

fn canonical_fence(line: &str) -> Result<String, Error> {
    let trimmed = line.trim_start();
    let rest = &trimmed[3..];
    let lang = rest.trim();
    if lang.is_empty() {
        return Ok("```".to_string());
    }
    if !lang
        .chars()
        .all(|c| c.is_ascii_alphanumeric() || matches!(c, '-' | '_' | '+' | '.'))
    {
        return Err(Error::Cbor("invalid code fence language"));
    }
    Ok(format!("```{}", lang))
}

fn canonicalize_heading(line: &str) -> Result<Option<String>, Error> {
    if !line.starts_with('#') {
        return Ok(None);
    }
    let mut level = 0usize;
    for ch in line.chars() {
        if ch == '#' {
            level += 1;
        } else {
            break;
        }
    }
    if level == 0 || level > MAX_HEADING_LEVEL {
        return Err(Error::Cbor("invalid heading"));
    }
    let rest = line[level..].trim_start();
    if rest.is_empty() {
        return Err(Error::Cbor("empty heading"));
    }
    Ok(Some(format!("{} {}", "#".repeat(level), rest)))
}

fn canonicalize_blockquote(line: &str) -> Result<Option<String>, Error> {
    if !line.starts_with('>') {
        return Ok(None);
    }
    let rest = line[1..].trim_start();
    Ok(Some(format!("> {}", rest)))
}

fn canonicalize_hr(line: &str) -> Result<Option<String>, Error> {
    if line.trim() == "---" {
        return Ok(Some("---".to_string()));
    }
    Ok(None)
}

fn canonicalize_list(line: &str) -> Result<Option<String>, Error> {
    let (indent, rest) = split_indent(line);
    if indent > 3 {
        return Err(Error::Cbor("excessive indent"));
    }
    if rest.starts_with('-') {
        let after = rest[1..].trim_start();
        if after.is_empty() || rest.len() == 1 || !rest.as_bytes()[1].is_ascii_whitespace() {
            return Err(Error::Cbor("invalid list marker"));
        }
        validate_links(after)?;
        return Ok(Some(format!("{}- {}", " ".repeat(indent), after)));
    }
    if rest.starts_with('*') || rest.starts_with('+') {
        return Err(Error::Cbor("invalid list marker"));
    }
    if let Some((digits, after_dot)) = rest.split_once('.') {
        if digits.chars().all(|c| c.is_ascii_digit()) {
            if digits != "1" {
                return Err(Error::Cbor("ordered list must use 1."));
            }
            let after = after_dot.trim_start();
            if after.is_empty()
                || !after_dot
                    .as_bytes()
                    .get(0)
                    .map(|b| b.is_ascii_whitespace())
                    .unwrap_or(false)
            {
                return Err(Error::Cbor("invalid ordered list"));
            }
            validate_links(after)?;
            return Ok(Some(format!("{}1. {}", " ".repeat(indent), after)));
        }
    }
    Ok(None)
}

fn split_indent(line: &str) -> (usize, &str) {
    let mut count = 0usize;
    for ch in line.chars() {
        if ch == ' ' {
            count += 1;
        } else {
            break;
        }
    }
    (count, &line[count..])
}

fn validate_links(line: &str) -> Result<(), Error> {
    let mut idx = 0usize;
    while let Some(start) = line[idx..].find("](") {
        let url_start = idx + start + 2;
        let rest = &line[url_start..];
        let end = rest.find(')').ok_or(Error::Cbor("invalid link"))?;
        let url = rest[..end].trim();
        let scheme_end = url.find(':').ok_or(Error::Cbor("invalid link"))?;
        let scheme = &url[..scheme_end].to_lowercase();
        if scheme != "https" && scheme != "agentnet" && scheme != "did" {
            return Err(Error::Cbor("invalid link scheme"));
        }
        idx = url_start + end + 1;
    }
    Ok(())
}

fn contains_html(line: &str) -> bool {
    let bytes = line.as_bytes();
    for i in 0..bytes.len() {
        if bytes[i] == b'<' {
            if let Some(next) = bytes.get(i + 1) {
                let ch = *next as char;
                if ch.is_ascii_alphabetic() || ch == '/' || ch == '!' {
                    return true;
                }
            }
        }
    }
    false
}

fn looks_like_table_separator(line: &str) -> bool {
    if !line.contains('|') {
        return false;
    }
    let trimmed = line.trim();
    if trimmed.is_empty() {
        return false;
    }
    for ch in trimmed.chars() {
        if ch != '|' && ch != '-' && ch != ':' && ch != ' ' {
            return false;
        }
    }
    true
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde::Deserialize;
    use std::fs;

    #[derive(Deserialize)]
    struct MarkdownTests {
        cases: Vec<MarkdownCase>,
    }

    #[derive(Deserialize)]
    struct MarkdownCase {
        id: String,
        input: String,
        canonical: String,
        valid: bool,
    }

    #[test]
    fn markdown_profile_vectors() {
        let data = fs::read_to_string("../../../../spec/agentnet-markdown-tests-v0.1.json")
            .expect("read markdown tests");
        let tests: MarkdownTests = serde_json::from_str(&data).expect("parse markdown tests");

        for case in tests.cases {
            let canonicalized = canonicalize_markdown_profile(&case.input);
            match canonicalized {
                Ok(val) => {
                    assert_eq!(val, case.canonical, "{} canonical mismatch", case.id);
                }
                Err(_) => {
                    assert!(
                        case.canonical.is_empty(),
                        "{} should not canonicalize",
                        case.id
                    );
                }
            }
            let valid = validate_markdown_profile(&case.input).is_ok();
            assert_eq!(valid, case.valid, "{} validity mismatch", case.id);
        }
    }
}
